src/cunumeric/index/choose_template.inl

```diff
@@ -0,0 +1,67 @@
+/* Copyright 2021 NVIDIA Corporation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+#include "cunumeric/pitches.h"
+
+namespace cunumeric {
+
+using namespace Legion;
+using namespace legate;
+
+template <VariantKind KIND, LegateTypeCode CODE, int DIM>
+struct ChooseImplBody;
+
+template <VariantKind KIND>
+struct ChooseImpl {
+  template <LegateTypeCode CODE, int DIM>
+  void operator()(ChooseArgs& args) const
+  {
+    using VAL = legate_type_of<CODE>;
+    auto rect = args.out.shape<DIM>();
+
+    Pitches<DIM - 1> pitches;
+    size_t volume = pitches.flatten(rect);
+    if (volume == 0) return;
+
+    auto out       = args.out.write_accessor<VAL, DIM>(rect);
+    auto index_arr = args.inputs[0].read_accessor<int, DIM>(rect);
+
+#ifndef LEGION_BOUNDS_CHECKS
+    // Check to see if this is dense or not
+    bool dense =
+      index_arr.accessor.is_dense_row_major(rect) && out.accessor.is_dense_row_major(rect);
+#else
+    // No dense execution if we're doing bounds checks
+    bool dense = false;
+#endif
+    std::vector<AccessorRO<VAL, DIM>> choices;
+    auto rect_c = args.inputs[1].shape<DIM>();
```

Comment:
Use each Store's `shape()` specifically in operations involving that Store, including taking an accessor and testing dense-ness. Don't use the shape of one argument to operate on another (like you do here by using the first choice array's shape on operations involving the other choice arrays).

At this stage we only have access to the shape of the sub-tiles this point task is operating on. The shape of the sub-tile for each Store should ideally be the same, since we are aligning the partitions, however this is unfortunately not the case. If an argument has been broadcasted on one dimension, the shape of the sub-tile for that argument on that dimension will not be correct.

Say we are running this operation on 4 processors:

```
x = ones((6,1)) + zeros((1,4))
```

Then the shape of the output will be `(6,4)`, and say it's split in 4 sub-tiles of shape `(3,2)`.

Within each point task, the output (which doesn't get broadcasted) will indeed span a rect of extent `(3,2)`. However, when asked for the shape of the rhs1 sub-tile, legate will report it as `(3,4)` -- the number for the broadcasted dimension is wrong (more precisely, the number reported for that dimension will be the size of the full array on that dimension). Similarly, the shape of the rhs2 sub-tile will be reported as `(6,2)`.